[mono] new Sdk that selects mono runtime components

lambdageek opened the pull request:
Related to #54417 

The contract is:
* The runtime defines `_MonoRuntimeAvailableComponents` which is an item list like `"hot_reload;diagnostic_tracing"`
* The workoad defines `_MonoComponent` which is some subset of the items that makes sense in the current configuration of the user app
* The workload calls the `_MonoSelectRuntimeComponents` target
  * The target returns `_MonoRuntimeSelectedComponents` (and `_MonoRuntimeSelectedStubComponents` if static linking) which are the _names_ of the comopnents that will be available (or stubbed out, if static linking)
  * The target returns `_MonoRuntimeComponentLink` which is a list of _filenames_ (relative to the runtime pack `native/` subdirectory) of the libraries that must be either copied (for dynamic loading) or statically linked (for static linking) with the runtime.  The items also have some metadata such as `IsStub`, `ComponentName` and `Linking`
  * The target also returns a `_MonoRuntimeComponentDontLink` list of filenames to exclude from linking for workloads which would rather take everything from `native/` and just exclude what we tell them to exclude.

lambdageek committed:
Work in progress: new Sdk that selects mono runtime components

lambdageek commented:
/cc @steveisok @akoeplinger  I think this isn't the right way - I really need the component-manifest.props file to be different for each platform. It's just a simple declarative list of the components and how the runtime is configured.  I don't want to duplicate all the logic from mono.proj.  and I don't want to hardcode the list of components.  It would be best if this could go in some data/ folder in the runtime pack itself.

lambdageek committed:
Add props and targets description to the components doc

lambdageek commented:
I wonder if we should just have a platform-specific Sdk nuget as part of the workload.  (`Microsoft.NET.Runtime.Platform.<RID>.Sdk` - or something like that)  Some place to stash RID-specific msbuild stuff for each target. I'm not sure how well that fits into the workloads design, though

lambdageek committed:
condition the _MonoRuntimeAvailableComponents by RuntimeIdentifier

lambdageek committed:
[cmake] Write a component-manifest.props file during build

If we're not building a mono aot cross compiler, generate a
component-manifest.props file in artifacts/obj/mono/<host>/ that indicates if
the host will use static or dynamic components, and a list of the available
components, and the properties for constructing their names.

lambdageek committed:
Build Microsoft.NETCore.App.Runtime.Mono.<RID>.Sdk shared framework nuget

It seems to also generate a symbols nuget.  And in the nuget there's a
tools/mono-sdk-what-is-this.deps.json file from the
SharedFrameworkHostFileNameOverride property.  It would be nice to exclude that
stuff.

lambdageek committed:
put the compoonent-manifest.targets into the Sdk

lambdageek committed:
delete WIP in mono/nuget/

lambdageek commented:
@akoeplinger @directhex @steveisok Ok, I am using the SharedFramework SDK to make a new `Microsoft.NETCore.App.Runtime.Mono.<RID>.Sdk` nuget - it seems to be working at least for `iossimulator-x64`.

There are some gross cmake hacks in here, too, because I don't really know cmake.

Could you take a look at what's here so far.

Next steps: update the workload.  Figure out some way to test this. 

(I tested that the msbuild target works by copy-pasting to a dummy project and setting the appropriate props and looking that the properties are getting set correctly.)

lambdageek committed:
fixup static component names in component-manifest.targets

lambdageek committed:
delete fixed fixme

lambdageek committed:
add missing $

lambdageek committed:
fix whitespace

lambdageek committed:
[cmake] switch to configure_file instead of file(CONFIGURE)

lambdageek committed:
add missing trailing slashes in .props.in file

lambdageek committed:
Add new Sdk packs to the workload manifest

lambdageek commented:
Ok, added a workload manifest.

One thing that is surprising to me is that we have apparently an iossimulator-x86 runtime pack?  Is that right?  I didn't add the workload manifest for the Sdk for it.  We need one, right?

lambdageek commented:
Is this actually going to work?  because of the indirection of going through pack IDs, only a build targeting, let's say... android-arm64 will load the `Microsoft.NETCore.App.Runtime.Mono.android-arm64.Sdk` nuget's Sdk.props and from there the component-manifest.props, right?  And so that build will get the mono components for android-arm64.

steveisok commented:
> Is this actually going to work? because of the indirection of going through pack IDs, only a build targeting, let's say... android-arm64 will load the `Microsoft.NETCore.App.Runtime.Mono.android-arm64.Sdk` nuget's Sdk.props and from there the component-manifest.props, right? And so that build will get the mono components for android-arm64.

You need to add the right imports in https://github.com/dotnet/runtime/blob/4615e06d531d10cb0c4347173718ee30620fda9a/src/mono/nuget/Microsoft.NET.Workload.Mono.Toolchain.Manifest/WorkloadManifest.targets

lambdageek commented:
> > Is this actually going to work? because of the indirection of going through pack IDs, only a build targeting, let's say... android-arm64 will load the `Microsoft.NETCore.App.Runtime.Mono.android-arm64.Sdk` nuget's Sdk.props and from there the component-manifest.props, right? And so that build will get the mono components for android-arm64.
> 
> You need to add the right imports in https://github.com/dotnet/runtime/blob/4615e06d531d10cb0c4347173718ee30620fda9a/src/mono/nuget/Microsoft.NET.Workload.Mono.Toolchain.Manifest/WorkloadManifest.targets

@steveisok Can I just do it like this?  is TargetArchitecture defined?

```
    <ImportGroup>
        <Import Project="Sdk.props" Sdk="Microsoft.NETCore.App.Runtime.$(TargetPlatformIdentifier)-$(TargetArchitecture).Sdk" />
        <Import Project="Sdk.targets" Sdk="Microsoft.NETCore.App.Runtime.$(TargetPlatformIdentifier)-$(TargetArchitecture).Sdk" />
   </ImportGroup>
```

lambdageek commented:
Ok, so the plan here is twofold:

1. Make the mono-component.props include the host RID in the metadata.  So we can include Sdk.props from multiple architectures without conflict.
2. Make a new nuget that just has the `mono-components.targets` in it that has the single target that is RID-independent.  Update the target to pick from the components for the current target only.
3. Update the workload manifest for each workload to bring in all the props and the single target

marek-safar commented:
@steveisok would be better to extract the plan to separate issue for tracking?

lambdageek commented:
I updated the related issue https://github.com/dotnet/runtime/issues/54417 with the notes from here.  I also plan to update the design doc as part of this PR.

lewing commented:
I think this could be simplified a lot if we handled it the way the Cross runtimes do?

steveisok commented:
> I think this could be simplified a lot if we handled it the way the Cross runtimes do? - which sound like roughly the plan from [#54432 (comment)](https://github.com/dotnet/runtime/pull/54432#issuecomment-867060652)

Yep, the difference is that we also want to include a .targets file.  We can only import that once.

lewing commented:
> > I think this could be simplified a lot if we handled it the way the Cross runtimes do? - which sound like roughly the plan from [#54432 (comment)](https://github.com/dotnet/runtime/pull/54432#issuecomment-867060652)
> 
> Yep, the difference is that we also want to include a .targets file. We can only import that once, so it has to be in another pack.

I think we could probably combine at least the runtime config and this into a single pack with targets that we enable appropriately.  There isn't much need to have everything in a separate pack is there?

steveisok commented:
No, there's not a real need to have them in a separate pack.

lambdageek commented:
>  think we could probably combine at least the runtime config and this into a single pack with targets that we enable appropriately. There isn't much need to have everything in a separate pack is there?

Yea I think we should have a single Mono Targeting SDK nuget that has all tasks and targets that are portable across all host and target architectures.  (At the moment it's the runtime config parser and this component selector)

lambdageek committed:
rework component-manifest.targets to use ItemGroups; move to new SDK

lambdageek committed:
Rename shared framework to Microsoft.NETCore.App.Runtime.Mono.<RID>.Props.Sdk

And only include component-manifest.props, not the targets

lambdageek committed:
Update manifest to include the new Props.Sdk and MonoTargets.Sdk

lambdageek commented:
Could I get someone to take a quick look at what I have so far before I move the runtimeconfig task.

There's now a `Microsoft.NETCore.App.Runtime.Mono.<RID>.Props.Sdk` that has the `component-manifest.props` per-RID, and a `Microsoft.NET.Runtime.MonoTargets.Sdk` which has the `component-manifest.targets`.

I updated the workload manifest so that it brings in all the props for all architectures for a particular OS.  So Android brings in `android-x86, android-x64, android-arm, android-arm64` props, for example.

One **question** I noticed in `src/mono/nuget/mono-packages.proj` that we only build the workload manifest when we `$(TargetsMobile)`, but not for Wasm.  Is that a problem?  Or maybe it doesn't matter because as long as the workload manifest gets built once that's good enough?

https://github.com/dotnet/runtime/blob/e129e327b60d49df7770333a3ced94e5085cf43d/src/mono/nuget/mono-packages.proj#L16-L17

marek-safar commented on src/mono/nuget/Microsoft.NET.Workload.Mono.Toolchain.Manifest/WorkloadManifest.json.in:
```diff
@@ -114,6 +133,22 @@
       "kind": "framework",
       "version": "${PackageVersion}"
     },
+    "Microsoft.NETCore.App.Runtime.Mono.android-arm.Props.Sdk": {
```
I'm not sure about .Sdk suffix here, it seems to be confusing

steveisok commented:
> Could I get someone to take a quick look at what I have so far before I move the runtimeconfig task.
> One **question** I noticed in `src/mono/nuget/mono-packages.proj` that we only build the workload manifest when we `$(TargetsMobile)`, but not for Wasm. Is that a problem? Or maybe it doesn't matter because as long as the workload manifest gets built once that's good enough?

`$(TargetsMobile)` should include wasm https://github.com/dotnet/runtime/blob/fc4a42746d9cf65ad3db4074207069ab9c168bc9/Directory.Build.props#L20

What you have looks good to me structurally.

lambdageek commented:
> What you have looks good to me structurally.

Ok, I'm going to move the runtime config task into the targets sdk and update the design doc.

Then we should do one more review and check for spelling mistakes.  I wish we had some validator for these things other than building a full workload and trying to run msbuild

steveisok commented:
Thanks to @radical, once https://github.com/dotnet/runtime/pull/54451 gets going we can begin to have some workload validation.

steveisok commented:
> ok, let's talk about naming.

`Microsoft.NETCore.App.Runtime.Mono.android-arm.HostProps` is ok to me. 

What about `Microsoft.NETCore.App.Runtime.Mono.android-arm.Component.BuildProps`?

I'm bad with names.

lambdageek committed:
Move RuntimeConfigParserTask into Microsoft.NET.Runtime.MonoTargets.Sdk

Consolidate all platform-independent tasks and targets into a single Sdk

lambdageek committed:
Add iossimulator-x86 props

lambdageek committed:
update components design doc

lambdageek commented:
> > ok, let's talk about naming.

> What about `Microsoft.NETCore.App.Runtime.Mono.android-arm.Component.BuildProps`?

I would prefer not too put `Component` in the name - we should probably use a single pack for all the runtime-identifier dependent properties (excluding cross compiler props which probably have a more complicated matrix).  Otherwise it's silly to have all the nuget overhead to essentially just ship one file per pack per platform.

lewing commented:
Do we really need an extra props pack? Can we not stick a props file in the appropriate runtime pack?

lambdageek commented:
> Do we really need an extra props pack? Can we not stick a props file in the appropriate runtime pack? 

I would also really like to understand the reasons for why this isn't ok.    

> We keep running into the situation where it would be nice to have more information about the contents of the runtime pack, it seems like we should solve that.

Absolutely.

> #48016

lambdageek commented:
Hmm... I think I need to actually add a `@(_MonoRuntimeComponentDontLink)` output from the components target - I think the actual way that the workloads will use this is to pass everything in `native/` to the linker (static) or app bundler (dynamic) _execept_ the disabled/stubbed out components.

@rolfbjarne @grendello @lateralusX @jonathanpeppers I'd love some feedback about the design (see the components.md).

jonathanpeppers commented on docs/design/mono/components.md:
```diff
@@ -265,16 +265,37 @@ To implement `feature_X` as a component.  Carry out the following steps:
+includes the runtime pack and additional tasks, properties and targets that
+enable the workload to construct a runtime for various scenarios.
+
+For the target RID, we expose:
+
+- `@(_MonoRuntimeComponentLinking)` set to either `'static'` or `'dynamic'` depending on whether the
+  current runtime pack for the current target includes runtime components as static archives or as
+  shared libraries, respectively.
+- `@(_MonoRuntimeComponentSharedLibExt)` and `@(_MonoRuntimeComponentStaticLibExt)` set to the file
+  extension of the runtime components for the current target (ie, `'.a', '.so', '.dylib'` etc).
+- `@(_MonoRuntimeAvailableComponents)` a list of component names without the `lib` prefix (if any)
+  or file extensions.  For example: `'hot_reload; diagnostics_tracing'`.
```
In this doc, could you include some example `.xml` snippets of what the new `<ItemGroup>`'s come out to be?

jonathanpeppers commented on src/mono/mono/component/component-manifest.props.in:
```diff
@@ -0,0 +1,10 @@
+    <_MonoRuntimeComponentStaticLibExt Include="${TemplateMonoRuntimeComponentStaticLibExt}" RuntimeIdentifier="${TemplateRuntimeIdentifier}" />
+  </ItemGroup>
+  <ItemGroup>
+    <_MonoRuntimeAvailableComponents Include="${TemplateMonoRuntimeAvailableComponents}" RuntimeIdentifier="${TemplateRuntimeIdenfifier}" />
```
Think there might be a typo here:
```suggestion
    <_MonoRuntimeAvailableComponents Include="${TemplateMonoRuntimeAvailableComponents}" RuntimeIdentifier="${TemplateRuntimeIdentifier}" />
```

jonathanpeppers commented on src/mono/nuget/Microsoft.NET.Runtime.MonoTargets.Sdk/Sdk/component-manifest.targets:
```diff
@@ -0,0 +1,62 @@
+<Project>
+  <Target Name="_MonoComputeAvailableComponentDefinitions">
+    <!-- Input: _MonoRuntimeAvailableComponents item list: the components provided by the runtime -->
+    <!-- Input: _MonoRuntimeComponentLinking property: either 'dynamic' or 'static'.  Are the components static or shared libraries. -->
```
What is intended to set `$(_MonoRuntimeComponentCurrentLinking)`? Is the Android workload going to default this to `dynamic` and iOS might have some other way to select it?

lambdageek replied:
> What is intended to set `$(_MonoRuntimeComponentCurrentLinking)`? Is the Android workload going to default this to `dynamic` and iOS might have some other way to select it?

Actually the runtime pack gets to choose, not the workload.

I should fixup this part of the doc - `@(_MonoRuntimeAvailableComponents)` and `@(_MonoRuntimeComponentLinking)` aren't really inputs to the target that the workload authors are expected to set.

grendello commented:
> Hmm... I think I need to actually add a `@(_MonoRuntimeComponentDontLink)` output from the components target - I think the actual way that the workloads will use this is to pass everything in `native/` to the linker (static) or app bundler (dynamic) _execept_ the disabled/stubbed out components.
> 

I also have a half-baked thought that it could be nice to know if the component is meant to be used for `debug,` `release` or both configurations. For instance, `hot_reload` is clearly a debug-only affair, `dotnet_tracing` is for both and there might also be components in the future that will be useful/usable only for one of the configurations.

lambdageek committed:
Fix typo

lambdageek committed:
improve docs

lambdageek committed:
Add _MonoRuntimeComponentDontLink target output

lewing commented on src/mono/nuget/Microsoft.NET.Workload.Mono.Toolchain.Manifest/WorkloadManifest.targets:
```diff
@@ -42,5 +53,8 @@
+        <Import Project="Sdk.props" Sdk="Microsoft.NETCore.App.Runtime.browser-wasm.Props.Sdk" />
     </ImportGroup>
+
+    <Import Project="Sdk.targets" Sdk="Microsoft.NET.Runtime.MonoTargets.Sdk" />
```
You need a condition here or these targets will be imported *everywhere*, this file is essentially part of the .NET SDK.

lambdageek replied:
What would the condition be?  We want this used everywhere, don't we?

steveisok replied:
Move it into the mportGroup for each of the platforms we support.  Otherwise, it will be used in places where we aren't using mono.

lewing commented:
Even if we don't solve the overal

> Hmm... I think I need to actually add a `@(_MonoRuntimeComponentDontLink)` output from the components target - I think the actual way that the workloads will use this is to pass everything in `native/` to the linker (static) or app bundler (dynamic) _execept_ the disabled/stubbed out components.
> 
> @rolfbjarne @grendello @lateralusX @jonathanpeppers I'd love some feedback about the design (see the components.md).

We need to link stubs in some cases right?

lambdageek commented:
> We need to link stubs in some cases right?

Yea the target does that now for static linking - if you select `"hot_reload"` it will tell you "don't link the hot_reload stub; don't link the diagnostic tracing full component".

lewing commented:
> > Do we really need an extra props pack? Can we not stick a props file in the appropriate runtime pack?
> 
> I would also really like to understand the reasons for why this isn't ok.
> 
> > We keep running into the situation where it would be nice to have more information about the contents of the runtime pack, it seems like we should solve that.
> 
> Absolutely.
> 
> > #48016

https://github.com/dotnet/runtime/issues/48016#issuecomment-781764687 in particular.

lambdageek commented:
> stick the .props in the runtime pack and exclude it from RuntimeList.xml and make sure it gets loaded

Ok, sticking it into the runtime pack seems fine.  (I assume I can just add a `build/` directory in the runtime pack nuget).

Not sure how we will go about loading it.  Presumably the Mono workload WorkloadManifest.targets would do it?  What would it import?  Can I just do something like:

```xml
  <Import Project="../build/mono-component.props" Sdk="Microsoft.NETCore.App.Runtime.Mono.$(RuntimeIdentifier)">
```

steveisok commented:
> Not sure how we will go about loading it. Presumably the Mono workload WorkloadManifest.targets would do it? What would it import? Can I just do something like:
> 
> ```
>   <Import Project="../build/mono-component.props" Sdk="Microsoft.NETCore.App.Runtime.Mono.$(RuntimeIdentifier)">
> ```
No, you would need to do something like this (we need to test it):

```
<Import Project="%(ResolvedRuntimePack.PackageDirectory)\runtimes\$(RuntimeIdentifier)\build\mono-component.props" Sdk="Microsoft.NETCore.App.Runtime.Mono.$(RuntimeIdentifier)" />
```

lewing commented:
should be able to follow the pattern here https://github.com/dotnet/runtime/pull/53685

lewing commented:
We actually build that up so there may need to be more work but it looks like we should be able to or construct it

lewing commented:
%(ResolvedRuntimePack.PackageDirectory) should get you the path

lambdageek committed:
Drop component-manifest.props into runtime pack build/ directory

Remove from the Microsoft.NETCore.App.Mono.Props.Sdk workload nuget

lambdageek committed:
remove Microsoft.NETCore.App.Mono.Props.Sdk

lambdageek committed:
Import component-manifest.props from the runtime pack

Move the MonoTargets.Sdk import to each target platform that we support

lambdageek commented:
Ok, no more Props.Sdk.  Bundling the files in the runtime pack in a `build/` directory.  And importing it from WorkloadManifest.targets using the ResolvedRuntimePack.   Also imports in the WorkloadManifest are now only done for RIDs we support.

rolfbjarne commented on docs/design/mono/components.md:
```diff
@@ -265,16 +265,66 @@ To implement `feature_X` as a component.  Carry out the following steps:
+    of the components that were (resp, were not) selected.  For example `'hot_reload;
+    diagnostics_tracing'`.  Each item has two metadata properties `ComponentLib` and
+    `ComponentStubLib` (which may be empty) that specify the name of the static or dynamic library
+    of the compponent.  This is not the main output of the target, it's primarily for debugging.
```
```suggestion
    of the component.  This is not the main output of the target, it's primarily for debugging.
```

rolfbjarne commented on docs/design/mono/components.md:
```diff
@@ -265,16 +265,66 @@ To implement `feature_X` as a component.  Carry out the following steps:
+    `'dynamic'`).  This output should be used by the workloads when linking the app and runtime if
+    the workload uses an allow list of native libraries to link or bundle.
+  - output `@(_MonoRuntimeComponentDontLink)` a list of library names (relative to the `native/`
+    subdirectory of the rutnime pack) that should be excluded from the application bundle (for
```
```suggestion
    subdirectory of the runtime pack) that should be excluded from the application bundle (for
```

rolfbjarne commented on docs/design/mono/components.md:
````diff
@@ -265,16 +265,66 @@ To implement `feature_X` as a component.  Carry out the following steps:
+
+```xml
+  <Project>
+    <ItemGroup Condition="'$(Configuration)' == 'Debug'>
````
```suggestion
    <ItemGroup Condition="'$(Configuration)' == 'Debug'">
```

rolfbjarne commented on src/mono/nuget/Microsoft.NET.Runtime.MonoTargets.Sdk/Sdk/component-manifest.targets:
```diff
@@ -0,0 +1,94 @@
+         in _MonoComponent -->
+    <!-- Output: _MonoRuntimeSelectedStubComponents (static linking only) subset of
+         _MonoRuntimeComponentName items that were not in _MonoComponent -->
+    <!-- Ouptut: _MonoRuntimeUnSelectedComponents subset of _MonoRuntimeComponentName items that were not in _MonoComponent. That is, the components that are available but that weren't selected. -->
```
```suggestion
    <!-- Output: _MonoRuntimeUnSelectedComponents subset of _MonoRuntimeComponentName items that were not in _MonoComponent. That is, the components that are available but that weren't selected. -->
```

lambdageek committed:
Fix typos

Co-authored-by: Rolf Bjarne Kvinge <rolf@xamarin.com>

radical commented on src/mono/nuget/Microsoft.NET.Runtime.MonoTargets.Sdk/Sdk/component-manifest.targets:
```diff
@@ -0,0 +1,94 @@
+      <!-- take the intersection of the two item lists. -->
+      <!-- n.b. copies the metadata from _MonoRuntimeComponentName to _MonoRuntimeSelectedComponents -->
+      <_MonoRuntimeComponentNameForRid Include="@(_MonoRuntimeComponentName)" Condition="'%(RuntimeIdentifier)' == '$(RuntimeIdentifier)'" />
+      <_MonoRuntimeSelectedComponents Include="@(_MonoRuntimeComponentNameForRid)" Condition="'@(_MonoRuntimeComponentNameForRid)' == '@(_MonoComponent)' and '%(Identity)' != ''" />
```
I *think*, `'@(_MonoRuntimeComponentNameForRid)' == '@(_MonoComponent)'` would compare the stringified version of the two items, which would be order sensitive.

lambdageek replied:
`Condition="'@(A)' == '@(B)' and '%(Identity)' != ''"` ends up comparing every item in list A to every item in list B one by one.  The `%(Identity)` makes it go item by item.  (This is apparently a trick for doing set intersections in msbuild https://philcupp.wordpress.com/2009/10/12/msbuild-tricks-i-don%E2%80%99t-want-to-forget/)

radical commented on src/mono/nuget/Microsoft.NET.Runtime.MonoTargets.Sdk/Sdk/component-manifest.targets:
```diff
@@ -0,0 +1,94 @@
+        <ComponentName>%(_MonoRuntimeSelectedStubComponents.Identity)</ComponentName>
+      </_MonoRuntimeComponentLink>
+      <_MonoRuntimeComponentDontLink Include="%(_MonoRuntimeUnSelectedStubComponents.ComponentStubLib)">
+	<Linking>$(_MonoRuntimeComponentCurrentLinking)</Linking>
+	<IsStub>true</IsStub>
+	<ComponentName>%(_MonoRuntimeUnSelectedStubComponents.Identity)</ComponentName>
```
tabs -> spaces

radical commented on src/mono/nuget/Microsoft.NET.Runtime.MonoTargets.Sdk/Sdk/component-manifest.targets:
```diff
@@ -0,0 +1,94 @@
+      </_MonoRuntimeComponentDontLink>
+    </ItemGroup>
+
+    <Error Condition="'$(_MonoRuntimeComponentCurrentLinking)' != 'dynamic' and '$(_MonoRuntimeComponentCurrentLinking)' != 'static'" Text="_MonoRuntimeComponentCurrentLinking is not dynamic or static.  It is '$(_MonoRuntimeComponentCurrentLinking)'.  RuntimeIdentifier is '$(RuntimeIdentifier)'." />
```
This check should be done earlier.

radical commented on src/mono/nuget/Microsoft.NET.Workload.Mono.Toolchain.Manifest/WorkloadManifest.targets:
```diff
@@ -10,7 +10,11 @@
-    <Import Condition="'$(TargetPlatformIdentifier)' == 'android'" Project="Sdk.props" Sdk="Microsoft.NET.Runtime.RuntimeConfigParser.Task" />
+    <ImportGroup Condition="'$(TargetPlatformIdentifier)' == 'android'">
+        <Import Project="Sdk.props" Sdk="Microsoft.NET.Runtime.MonoTargets.Sdk" />
+	<Import Project="%(ResolvedRuntimePack.PackageDirectory)\runtimes\$(RuntimeIdentifier)\build\component-manifest.props"/>
+	<Import Project="Sdk.targets" Sdk="Microsoft.NET.Runtime.MonoTargets.Sdk" />
```
tabs -> spaces, here and in other places in this file

radical commented on src/mono/nuget/Microsoft.NET.Runtime.MonoTargets.Sdk/Sdk/component-manifest.targets:
```diff
@@ -0,0 +1,94 @@
+         _MonoRuntimeComponentName items that were not in _MonoComponent -->
+    <!-- Output: _MonoRuntimeUnSelectedComponents subset of _MonoRuntimeComponentName items that were not in _MonoComponent. That is, the components that are available but that weren't selected. -->
+    <!-- Output: _MonoRuntimeUnSelectedStubComponents (static linking only) subset of _MonoRuntimeComponentName items that were in _MonoComponent.  That is, components that were selected - so their stubs should not be linked. -->
+    <!-- Output: _MonoRuntimeComponentLink item list of the filenames to link (relative paths
+         relative to the native/ directory of the runtime pack). For dynamic linking, this is only
+         the enabled components.  For static linking, it also includes stubs for the non-selected
+         components -->
+    <!-- Output: _MonoRuntimeComponentDontLink item list of the filename to exclude from linking
+         (relative paths relative to the native/ directory of the runtime pack). For static linking,
+         this is stubs or components that must not be passed to the native linker. For dynamic
+         linking, this is copmonent shared libraries that must not be bundled with the app
+         -->
```
Should these two be combined into one item, and differentiated by some metadata?
And same question about `_MonoRuntime*Selected*Components`.

lambdageek replied:
No, they're completely unrelated.  One is for workloads that want to maintain a whitelist of stuff from `native/` that they will use.  The other is for workloads that just want to grab `native/*.a` and want an exclude list to tell them what to leave out.

radical commented on src/mono/nuget/Microsoft.NET.Workload.Mono.Toolchain.Manifest/WorkloadManifest.targets:
```diff
@@ -19,28 +23,44 @@
 
     <ImportGroup Condition="'$(RuntimeIdentifier)' == 'browser-wasm' and '$(UsingBrowserRuntimeWorkload)' == 'true'">
+        <Import Project="Sdk.props" Sdk="Microsoft.NET.Runtime.MonoTargets.Sdk" />
+	<Import Project="%(ResolvedRuntimePack.PackageDirectory)\runtimes\$(RuntimeIdentifier)\build\component-manifest.props"/>
```
Does `%(ResolvedRuntimePack.PackageDirectory)` work here? IIUC, batching shouldn't work here at all. And AFAIK, `@(ResolvedRuntimePack)` is set by `ResolveFrameworkReferences`. But I could be wrong. Does it work for you?

steveisok replied:
I wasn't 100% sure either and I think we need to test to find out

lewing replied:
I tested it in a target at which point it would be resolved I think we may need a couple different plan.

lambdageek replied:
Pinged some MSBuild folks.  One thing I coudl do is use the `<MSBuild/>` task in my MonoTargets target to read the manifest, but since there's a couple different things in there, it's actually a little bit easier/faster/cleaner to write a little Task that will read a component manifest json blob and get everything from there.  So I'll hack that up (probably tomorrow)

radical replied:
That task will be useful for wasm build targets too!

lambdageek committed:
Add JsonToItemsTaskFactory

lambdageek committed:
fix whitespace

lambdageek committed:
Apply suggestions from code review

Co-authored-by: Ankit Jain <radical@gmail.com>

lambdageek committed:
Do some validation earlier in _MonoComputeAvailableComponentDefinitions

lambdageek committed:
Read component-manifest.json using the JsonToItemsTaskFactory

and bundle it in Microsoft.NET.Runtime.MonoTargets.Sdk

lambdageek committed:
remove ResolvedRuntimePack import from WorkloadManifest.targets

it's too early, and we have the JsonToItemsTaskFactory now to read the manifest

lambdageek committed:
Generate component-manifest.json in CMakeLists.txt

lambdageek commented:
Ok, this is truly baroque now.

* The cmake file generates `mono-component.json` which goes into each runtime pack.
* The MonoTargets SDK uses the new `JsonToItemsTaskFactory` MSBuild task to read the json into items in a target that assumes `%(ResolvedRuntimePack.PackageDir)` is sensible.
* After that, `_MonoSelectRuntimeComponents` can run and populate the Link and DontLink item lists with the names of the shared or static libraries that the workload should use for including/excluding components.

Assuming there are no typos anywhere it might even work.

lambdageek commented on src/mono/nuget/Microsoft.NET.Runtime.MonoTargets.Sdk/Sdk/component-manifest.targets:
```diff
@@ -0,0 +1,94 @@
+<Project>
+  <Target Name="_MonoComputeAvailableComponentDefinitions">
+    <!-- Input: _MonoRuntimeAvailableComponents item list: the components provided by the runtime -->
```
Not sure what kind of validation you would like to see here.

The item list is coming from the json blob now.

lambdageek commented on src/mono/nuget/Microsoft.NET.Workload.Mono.Toolchain.Manifest/WorkloadManifest.targets:
```diff
@@ -10,43 +10,51 @@
-    <ImportGroup Condition="'$(TargetPlatformIdentifier)' == 'ios'">
-        <Import Project="Sdk.props" Sdk="Microsoft.NET.Runtime.RuntimeConfigParser.Task" />
-    </ImportGroup>
     <ImportGroup Condition="'$(TargetPlatformIdentifier)' == 'ios' and $([MSBuild]::IsOSPlatform('osx'))">
+        <Import Project="Sdk.props" Sdk="Microsoft.NET.Runtime.MonoTargets.Sdk" />
```
Should this be included if the TPI is 'ios' even if we're not on a Mac, like RuntimeConfigParser.Task was before?  (and same for catalyst and tvos)

steveisok replied:
Does the ios offline build on windows only build the managed parts?  If so, then it does not need to be included beyond ios.

lambdageek committed:
Fix some copy-paste nits

lambdageek commented:
@steveisok @lewing @radical I don't think there is any major outstanding work remaining on this PR.

steveisok commented on src/installer/pkg/sfx/Microsoft.NETCore.App/Microsoft.NETCore.App.Runtime.props:
```diff
@@ -70,6 +70,13 @@
       </RuntimeFiles>
 
+      <!-- FIXME: Can this condition be '$(RuntimeFlavor)' == 'mono' and not TargetsMobile? We will want this for desktop mono runtime packs too -->
+      <RuntimeFiles Condition="'$(TargetsMobile)' == 'true'"
```
You should be able to rely on '$(RuntimeFlavor)' == 'mono' for all our configurations

steveisok reviewed (approved):
LGTM - nice work!

radical commented on src/tasks/JsonToItemsTaskFactory/JsonToItemsTaskFactory.cs:
```diff
@@ -0,0 +1,379 @@
+            {
+                if (jsonFilePath == null)
+                {
+                    Log.LogError("no JsonFilePath specified");
```
```suggestion
                    Log.LogError($"no {nameof(JsonFilePath)} specified");
```

radical commented on src/tasks/JsonToItemsTaskFactory/JsonToItemsTaskFactory.cs:
```diff
@@ -0,0 +1,379 @@
+                    json = JsonSerializer.DeserializeAsync<JsonModelRoot>(file, JsonOptions).AsTask().Result;
+                    if (json == null)
+                    {
+                        Log.LogError ("Failed to deserialize Json file");
```
Might be useful to have the json file path here

radical commented on src/tasks/JsonToItemsTaskFactory/JsonToItemsTaskFactory.cs:
```diff
@@ -0,0 +1,379 @@
+                }
+            }
+
+            internal void LogParsedJson (JsonModelRoot json)
```
Maybe the log messages from this should use `MessageImportance.Low`, instead of the default `MessageImportance.Normal`? and same for other debug messages

lewing reviewed (commented):
nit: could we s/component-manifest/RuntimeComponentManifest/ or something, adding yet another naming format is painful

lambdageek committed:
Use RuntimeFlavor==mono for runtime pack build directory

instead of TargetsMobile.  We want the build files (mono-components.json)
in every mono runtime pack, not just on mobile targets

radical commented on src/tasks/JsonToItemsTaskFactory/JsonToItemsTaskFactory.cs:
```diff
@@ -0,0 +1,379 @@
+        public class JsonModelRoot
+        {
+            [JsonConverter(typeof(CaseInsensitiveDictionaryConverter))]
+            public Dictionary<string, string>?  Properties {get; set;}
```
```suggestion
            public Dictionary<string, string>? Properties {get; set;}
```

radical commented on src/tasks/JsonToItemsTaskFactory/JsonToItemsTaskFactory.cs:
```diff
@@ -0,0 +1,379 @@
+                    {
+                        foreach (var metadata in itemModel.Metadata)
+                        {
+                            if (string.Equals(metadata.Key, "Identity", StringComparison.OrdinalIgnoreCase))
```
should we emit a warning in this case, to catch an authoring error?

radical commented on src/tasks/JsonToItemsTaskFactory/JsonToItemsTaskFactory.cs:
```diff
@@ -0,0 +1,379 @@
+                for (int i = 0; i < itemModels.Length; i++)
+                {
+                    var itemModel = itemModels[i];
+                    var item = new TaskItem(itemModel.Identity);
```
Check for null?

radical commented on src/tasks/JsonToItemsTaskFactory/JsonToItemsTaskFactory.cs:
```diff
@@ -0,0 +1,379 @@
+
+        public class CaseInsensitiveDictionaryConverter : JsonConverter<Dictionary<string, string>>
+        {
+            public override Dictionary<string, string> Read (ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
```
Is this the method being overridden here? https://docs.microsoft.com/en-us/dotnet/api/system.text.json.serialization.jsonconverter-1.read?view=net-5.0#system-text-json-serialization-jsonconverter-1-read(system-text-json-utf8jsonreader-system-type-syst

Should the return type be nullable then?

lambdageek replied:
Apparently not. If I change the return type to `Dictionary<string,string>?`, I get (probably from the net472 build):

```
/Users/alklig/work/dotnet-runtime/runtime-bugs/src/tasks/JsonToItemsTaskFactory/JsonToItemsTaskFactory.cs(330,57): error CS8764: Nullability of return type doesn't match overridden member (possibly because of nullability attributes). [/Users/alklig/work/dotnet-runtime/runtime-bugs/src/tasks/JsonToItemsTaskFactory/JsonToItemsTaskFactory.csproj]
```

radical commented on src/tasks/JsonToItemsTaskFactory/JsonToItemsTaskFactory.cs:
```diff
@@ -0,0 +1,379 @@
+            {
+                var dict = JsonSerializer.Deserialize<Dictionary<string, string>>(ref reader, options);
+                if (dict == null)
+                    return null!;
```
Should the uses of JsonSerializer.Deserialize* also catch `JsonException`?

lambdageek replied:
I don't think so - we want it to propagate

radical replied:
Ok, then we should catch this up the chain somewhere, so we can log it as an error with the filename.

lambdageek committed:
rename component-manifest to RuntimeComponentManifest

radical commented on src/tasks/JsonToItemsTaskFactory/JsonToItemsTaskFactory.cs:
```diff
@@ -0,0 +1,379 @@
+                        if  (!idict.TryGetValue("Identity", out var identity) || identity == null)
+                            throw new Exception ("deserialized json dictionary item did not have an Identity metadata");
+                        item.Identity = identity;
+                        item.Metadata = idict;
```
Should the `Identity` entry be removed from `idict.Metadata`? We wouldn't need to special case it in the task then

lambdageek replied:
Did this and made `JsonModelItem.Identity` non-nullable, which improved things elsewhere. Thanks!

radical commented on src/tasks/JsonToItemsTaskFactory/JsonToItemsTaskFactory.cs:
```diff
@@ -0,0 +1,379 @@
+                        if (dict == null)
+                            return null!;
+                        var idict = new Dictionary<string, string> (dict, StringComparer.OrdinalIgnoreCase);
+                        if  (!idict.TryGetValue("Identity", out var identity) || identity == null)
```
Does this need to check for empty string too?

radical commented on src/tasks/JsonToItemsTaskFactory/JsonToItemsTaskFactory.csproj:
```diff
@@ -0,0 +1,32 @@
+    <PackageReference Include="Microsoft.Build.Framework" Version="$(RefOnlyMicrosoftBuildFrameworkVersion)" />
+    <PackageReference Include="Microsoft.Build.Tasks.Core" Version="$(RefOnlyMicrosoftBuildTasksCoreVersion)" />
+    <PackageReference Include="Microsoft.Build.Utilities.Core" Version="$(RefOnlyMicrosoftBuildUtilitiesCoreVersion)" />
+    <PackageReference Include="System.Reflection.Metadata" Version="5.0.0" />
```
Why `5.0.0`, and not `$(SystemReflectionMetadataVersion)`?

lambdageek committed:
fixup nullability annotations

lambdageek committed:
fix whitespace

lambdageek commented on src/tasks/JsonToItemsTaskFactory/JsonToItemsTaskFactory.cs:
```diff
@@ -0,0 +1,379 @@
+
+        public Type TaskType => typeof(JsonToItemsTask);
+
+        public bool Initialize (string taskName, IDictionary<string, TaskPropertyInfo> parameterGroup, string? taskBody, IBuildEngine taskFactoryLoggingHost)
```
Not sure. I'll experiment.  My feeling is that it won't be - `Initialize` isn't called anywhere in this project (it's called by the MSBuild engine), so the annotation won't help anywhere.  On the other hand we build this file for net472, so I'd need a local definition of `MemberNotNullAttribute` which is ugly.

radical replied:
Feel free to ignore!

lambdageek committed:
fix formatting

lambdageek committed:
Apply suggestions from code review

Co-authored-by: Ankit Jain <radical@gmail.com>

lambdageek committed:
Misc fixes to JsonToItemsTaskFactory

lambdageek committed:
Rename MonoRuntimeComponentManifestReadTask

from MonoRuntimeComponentsReadManifestTask

radical commented:
Don't wanna block you, but would it be possible to add a sample for this? In future we will be able to test this, at least on wasm. But for now, a sample would help validate it. Not sure if that makes sense though!

lambdageek committed:
undo nullability annotation

Build doesn't like it for some reason (probably net472)

lambdageek committed:
fix incorrect task parameter name

lambdageek commented:
> Don't wanna block you, but would it be possible to add a sample for this? In future we will be able to test this, at least on wasm. But for now, a sample would help validate it. Not sure if that makes sense though!

@radical 
It's hard to test the workloads end to end.  Here's a gist of what I use for testing (but as you noticed, I didn't test every time). Not sure where's a good place to check this in.  https://gist.github.com/lambdageek/7bb9672fa0a95bb232926aa7630962f4

lambdageek committed:
Remove Identity metadata from dictionary at json parsing time

Also improve nullability a bit by making the properties immutable

radical commented:
> It's hard to test the workloads end to end. Here's a gist of what I use for testing (but as you noticed, I didn't test every time). Not sure where's a good place to check this in. [gist.github.com/lambdageek/7bb9672fa0a95bb232926aa7630962f4](https://gist.github.com/lambdageek/7bb9672fa0a95bb232926aa7630962f4)

Makes sense. I have a PR open that will add some E2E(almost) testing in `runtime`, and I can add a test for this after that.

lambdageek committed:
Throw correct json deserializer exceptions

lambdageek committed:
Catch JsonException in an async function

We get nice error messages now like

```
src/mono/nuget/Microsoft.NET.Runtime.MonoTargets.Sdk/Sdk/RuntimeComponentManifest.targets(8,5): error : Failed to deserialize json from file 'artifacts/bin/mono/iOSSimulator.x64.Release/build/RuntimeComponentManifest.json', JSON Path: $.items._MonoRuntimeAvailableComponents[2], Line: 14, Position: 1 [component-manifest.sample.proj]
src/mono/nuget/Microsoft.NET.Runtime.MonoTargets.Sdk/Sdk/RuntimeComponentManifest.targets(8,5): error : JsonException: The JSON value could not be converted to System.Collections.Generic.Dictionary`2[System.String,System.String]. Path: $.identity | LineNumber: 0 | BytePositionInLine: 16. Path: $.items._MonoRuntimeAvailableComponents[2] | LineNumber: 14 | BytePositionInLine: 1. [component-manifest.sample.proj]
src/mono/nuget/Microsoft.NET.Runtime.MonoTargets.Sdk/Sdk/RuntimeComponentManifest.targets(8,5): error : InvalidOperationException: Cannot get the value of a token type 'Number' as a string. [component-manifest.sample.proj]
src/mono/nuget/Microsoft.NET.Runtime.MonoTargets.Sdk/Sdk/RuntimeComponentManifest.targets(8,5):
error :
[component-manifest.sample.proj]
```

radical commented on src/tasks/JsonToItemsTaskFactory/JsonToItemsTaskFactory.cs:
```diff
@@ -353,20 +370,20 @@ public override JsonModelItem Read(ref Utf8JsonReader reader, Type typeToConvert
                         return new JsonModelItem(identity, metadata: idict);
                     default:
-                        throw new Exception();
+                        throw new NotSupportedException();
```
token type in the message might be useful for debugging

lambdageek replied:
The JsonSerializer catches and decorates `NotSupportedException` with location and type info https://docs.microsoft.com/en-us/dotnet/standard/serialization/system-text-json-converters-how-to?pivots=dotnet-5-0#notsupportedexception

lambdageek committed:
fixup comments
